File: windows/EMWaver/Scripting/ScriptSourceTranspiler.cs
using System.Text;
using System.Text.RegularExpressions;

namespace EMWaver.Scripting;

internal static partial class ScriptSourceTranspiler
{
    internal static string Transpile(string source)
    {
        return new JsxParser(TranspileImports(source ?? string.Empty)).Transpile();
    }

    private static string TranspileImports(string source)
    {
        var lines = source.Split('\n');
        var output = new StringBuilder(source.Length + 128);
        var moduleIndex = 0;

        for (var i = 0; i < lines.Length; i++)
        {
            var line = lines[i].TrimEnd('\r');
            var transformed = TransformImportLine(line, moduleIndex);
            if (transformed != null)
            {
                moduleIndex += 1;
                output.Append(transformed);
            }
            else
            {
                output.Append(line);
            }
            if (i < lines.Length - 1)
            {
                output.Append('\n');
            }
        }

        return output.ToString();
    }

    private static string? TransformImportLine(string line, int moduleIndex)
    {
        var trimmed = line.Trim();
        if (!trimmed.StartsWith("import ", StringComparison.Ordinal))
        {
            return null;
        }

        var leading = LeadingWhitespace(line);
        var named = NamedImportRegex().Match(trimmed);
        if (named.Success)
        {
            var moduleVar = "__emw_mod_" + moduleIndex.ToString(System.Globalization.CultureInfo.InvariantCulture);
            var output = new StringBuilder()
                .Append(leading)
                .Append("var ")
                .Append(moduleVar)
                .Append(" = require(\"")
                .Append(named.Groups[2].Value)
                .Append("\");");

            foreach (var rawBinding in named.Groups[1].Value.Split(','))
            {
                var binding = rawBinding.Trim();
                if (binding.Length == 0) continue;
                var pieces = binding.Split((char[]?)null, StringSplitOptions.RemoveEmptyEntries);
                string exported;
                string local;
                if (pieces.Length == 1)
                {
                    exported = pieces[0];
                    local = pieces[0];
                }
                else if (pieces.Length == 3 && pieces[1] == "as")
                {
                    exported = pieces[0];
                    local = pieces[2];
                }
                else
                {
                    throw new InvalidOperationException("Unsupported import binding: " + binding);
                }
                output.Append(" var ").Append(local).Append(" = ").Append(moduleVar).Append('.').Append(exported).Append(';');
            }
            return output.ToString();
        }

        var namespaceImport = NamespaceImportRegex().Match(trimmed);
        if (namespaceImport.Success)
        {
            return $"{leading}var {namespaceImport.Groups[1].Value} = require(\"{namespaceImport.Groups[2].Value}\");";
        }

        var sideEffect = SideEffectImportRegex().Match(trimmed);
        if (sideEffect.Success)
        {
            return $"{leading}require(\"{sideEffect.Groups[1].Value}\");";
        }

        throw new InvalidOperationException("Unsupported import syntax: " + trimmed);
    }

    private static string LeadingWhitespace(string value)
    {
        var index = 0;
        while (index < value.Length && (value[index] == ' ' || value[index] == '\t'))
        {
            index += 1;
        }
        return value[..index];
    }

    [GeneratedRegex("^import\\s+\\{([^}]+)\\}\\s+from\\s+[\"']([^\"']+)[\"'];?\\s*$")]
    private static partial Regex NamedImportRegex();

    [GeneratedRegex("^import\\s+\\*\\s+as\\s+([A-Za-z_$][A-Za-z0-9_$]*)\\s+from\\s+[\"']([^\"']+)[\"'];?\\s*$")]
    private static partial Regex NamespaceImportRegex();

    [GeneratedRegex("^import\\s+[\"']([^\"']+)[\"'];?\\s*$")]
    private static partial Regex SideEffectImportRegex();

    private sealed class JsxParser
    {
        private readonly string _source;
        private int _index;

        internal JsxParser(string source)
        {
            _source = source;
        }

        internal string Transpile()
        {
            var output = new StringBuilder(_source.Length + 128);
            while (!IsAtEnd)
            {
                if (StartsLineComment) output.Append(ConsumeLineComment());
                else if (StartsBlockComment) output.Append(ConsumeBlockComment());
                else if (Current is '"' or '\'' or '`') output.Append(ConsumeQuotedString(Current));
                else if (Current == '<' && LooksLikeJsxElementStart(_index)) output.Append(ParseElement());
                else output.Append(Advance());
            }
            return output.ToString();
        }

        private string ParseElement()
        {
            Consume('<');
            var tag = ParseTagName();
            if (tag.Length == 0) throw new InvalidOperationException("unterminated JSX element");

            var attributes = new List<(string Name, string Value)>();
            var selfClosing = false;
            while (!IsAtEnd)
            {
                SkipWhitespace();
                if (StartsWith("/>"))
                {
                    _index += 2;
                    selfClosing = true;
                    break;
                }
                if (Current == '>')
                {
                    Advance();
                    break;
                }
                attributes.Add(ParseAttribute());
            }

            var children = new List<string>();
            var closed = selfClosing;
            while (!selfClosing && !IsAtEnd)
            {
                if (StartsWith("</"))
                {
                    _index += 2;
                    var closingTag = ParseTagName();
                    SkipWhitespace();
                    Consume('>');
                    if (closingTag != tag)
                    {
                        throw new InvalidOperationException($"mismatched JSX closing tag: expected </{tag}>, found </{closingTag}>");
                    }
                    closed = true;
                    break;
                }
                if (Current == '<' && LooksLikeJsxElementStart(_index))
                {
                    children.Add(ParseElement());
                    continue;
                }
                if (Current == '{')
                {
                    var expression = ParseBraceExpression();
                    if (!string.IsNullOrWhiteSpace(expression)) children.Add(expression);
                    continue;
                }
                var normalized = NormalizeText(ParseTextChild());
                if (normalized.Length > 0) children.Add(JsStringLiteral(normalized));
            }
            if (!closed) throw new InvalidOperationException($"unterminated JSX element <{tag}>");

            var args = new List<string> { TagReference(tag), MakeProps(attributes) };
            args.AddRange(children);
            return "JSX.h(" + string.Join(", ", args) + ")";
        }

        private (string Name, string Value) ParseAttribute()
        {
            var name = ParseAttributeName();
            if (name.Length == 0) throw new InvalidOperationException("unsupported JSX attribute near " + Current);
            SkipWhitespace();
            if (Current != '=') return (name, "true");
            Advance();
            SkipWhitespace();
            if (Current == '{') return (name, ParseBraceExpression());
            if (Current is '"' or '\'') return (name, ConsumeQuotedString(Current));
            var start = _index;
            while (!IsAtEnd && !char.IsWhiteSpace(Current) && Current != '>' && !StartsWith("/>")) Advance();
            return (name, _source[start.._index]);
        }

        private string ParseBraceExpression()
        {
            Consume('{');
            var start = _index;
            var depth = 1;
            while (!IsAtEnd)
            {
                if (StartsLineComment) { ConsumeLineComment(); continue; }
                if (StartsBlockComment) { ConsumeBlockComment(); continue; }
                if (Current is '"' or '\'' or '`') { ConsumeQuotedString(Current); continue; }
                var c = Advance();
                if (c == '{') depth += 1;
                else if (c == '}')
                {
                    depth -= 1;
                    if (depth == 0) return _source[start..(_index - 1)];
                }
            }
            throw new InvalidOperationException("unterminated JSX expression");
        }

        private string ParseTextChild()
        {
            var start = _index;
            while (!IsAtEnd && Current != '<' && Current != '{') Advance();
            return _source[start.._index];
        }

        private string ParseTagName()
        {
            var start = _index;
            while (!IsAtEnd && IsTagNameCharacter(Current)) Advance();
            return _source[start.._index];
        }

        private string ParseAttributeName()
        {
            var start = _index;
            while (!IsAtEnd && IsAttributeNameCharacter(Current)) Advance();
            return _source[start.._index];
        }

        private string MakeProps(List<(string Name, string Value)> attributes)
        {
            if (attributes.Count == 0) return "null";
            return "{ " + string.Join(", ", attributes.Select(a => PropertyKey(a.Name) + ": " + a.Value)) + " }";
        }

        private string PropertyKey(string name) => IsIdentifier(name) ? name : JsStringLiteral(name);
        private string TagReference(string tag) => IsIdentifier(tag) ? tag : JsStringLiteral(tag);
        private string NormalizeText(string text) => string.Join(" ", text.Split((char[]?)null, StringSplitOptions.RemoveEmptyEntries));

        private string JsStringLiteral(string value)
        {
            var output = new StringBuilder("\"");
            foreach (var c in value)
            {
                output.Append(c switch
                {
                    '"' => "\\\"",
                    '\\' => "\\\\",
                    '\n' => "\\n",
                    '\r' => "\\r",
                    '\t' => "\\t",
                    < ' ' => "\\u" + ((int)c).ToString("X4", System.Globalization.CultureInfo.InvariantCulture),
                    _ => c.ToString()
                });
            }
            output.Append('"');
            return output.ToString();
        }

        private bool IsAtEnd => _index >= _source.Length;
        private char Current => IsAtEnd ? '\0' : _source[_index];
        private char Peek => _index + 1 < _source.Length ? _source[_index + 1] : '\0';
        private char Advance() => _source[_index++];
        private bool StartsLineComment => Current == '/' && Peek == '/';
        private bool StartsBlockComment => Current == '/' && Peek == '*';
        private bool StartsWith(string text) => _source.StartsWith(text, _index, StringComparison.Ordinal);
        private void SkipWhitespace() { while (!IsAtEnd && char.IsWhiteSpace(Current)) Advance(); }
        private void Consume(char expected) { if (Current != expected) throw new InvalidOperationException("expected " + expected); Advance(); }

        private string ConsumeLineComment()
        {
            var start = _index;
            while (!IsAtEnd && Current != '\n') Advance();
            if (!IsAtEnd) Advance();
            return _source[start.._index];
        }

        private string ConsumeBlockComment()
        {
            var start = _index;
            _index += 2;
            while (!IsAtEnd)
            {
                if (Current == '*' && Peek == '/') { _index += 2; break; }
                Advance();
            }
            return _source[start.._index];
        }

        private string ConsumeQuotedString(char quote)
        {
            var start = _index;
            Advance();
            while (!IsAtEnd)
            {
                var c = Advance();
                if (c == '\\') { if (!IsAtEnd) Advance(); continue; }
                if (c == quote) break;
            }
            return _source[start.._index];
        }

        private bool LooksLikeJsxElementStart(int start)
        {
            if (start >= _source.Length || _source[start] != '<') return false;
            var cursor = start + 1;
            if (cursor >= _source.Length || !char.IsUpper(_source[cursor])) return false;
            while (cursor < _source.Length && IsTagNameCharacter(_source[cursor])) cursor += 1;
            while (cursor < _source.Length && char.IsWhiteSpace(_source[cursor])) cursor += 1;
            if (cursor >= _source.Length) return false;
            var c = _source[cursor];
            if (c == '>') return true;
            if (c == '/') return cursor + 1 < _source.Length && _source[cursor + 1] == '>';
            return IsAttributeNameStart(c);
        }

        private bool IsTagNameCharacter(char c) => char.IsLetterOrDigit(c) || c == '_' || c == '.';
        private bool IsAttributeNameCharacter(char c) => char.IsLetterOrDigit(c) || c == '_' || c == '-' || c == ':';
        private bool IsAttributeNameStart(char c) => char.IsLetter(c) || c == '_' || c == ':';
        private bool IsIdentifier(string value)
        {
            if (string.IsNullOrEmpty(value)) return false;
            if (!char.IsLetter(value[0]) && value[0] != '_' && value[0] != '$') return false;
            return value.Skip(1).All(c => char.IsLetterOrDigit(c) || c == '_' || c == '$');
        }
    }
}

File: android/app/src/main/java/com/emwaver/emwaverandroidapp/scripts/ScriptSourceTranspiler.java
/*
 * EMWaver
 * Copyright (c) 2026 Luis Marnoto
 * SPDX-License-Identifier: Apache-2.0
 */

package com.emwaver.emwaverandroidapp.scripts;

import java.util.ArrayList;
import java.util.List;
import java.util.Locale;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

final class ScriptSourceTranspiler {
    private static final Pattern NAMED_IMPORT = Pattern.compile("^import\\s+\\{([^}]+)\\}\\s+from\\s+[\"']([^\"']+)[\"'];?\\s*$");
    private static final Pattern NAMESPACE_IMPORT = Pattern.compile("^import\\s+\\*\\s+as\\s+([A-Za-z_$][A-Za-z0-9_$]*)\\s+from\\s+[\"']([^\"']+)[\"'];?\\s*$");
    private static final Pattern SIDE_EFFECT_IMPORT = Pattern.compile("^import\\s+[\"']([^\"']+)[\"'];?\\s*$");

    private ScriptSourceTranspiler() {}

    static String transpile(String source) {
        return new JsxParser(transpileImports(source != null ? source : "")).transpile();
    }

    private static String transpileImports(String source) {
        String[] lines = source.split("\\R", -1);
        StringBuilder out = new StringBuilder(source.length() + 128);
        int moduleIndex = 0;
        for (int i = 0; i < lines.length; i++) {
            String transformed = transformImportLine(lines[i], moduleIndex);
            if (transformed != null) {
                moduleIndex += 1;
                out.append(transformed);
            } else {
                out.append(lines[i]);
            }
            if (i < lines.length - 1) out.append('\n');
        }
        return out.toString();
    }

    private static String transformImportLine(String line, int moduleIndex) {
        String trimmed = line.trim();
        if (!trimmed.startsWith("import ")) return null;

        String leading = leadingWhitespace(line);
        Matcher named = NAMED_IMPORT.matcher(trimmed);
        if (named.matches()) {
            String moduleVar = "__emw_mod_" + moduleIndex;
            StringBuilder out = new StringBuilder(leading)
                    .append("var ")
                    .append(moduleVar)
                    .append(" = require(\"")
                    .append(named.group(2))
                    .append("\");");
            for (String binding : named.group(1).split(",")) {
                String part = binding.trim();
                if (part.isEmpty()) continue;
                String[] pieces = part.split("\\s+");
                String exported;
                String local;
                if (pieces.length == 1) {
                    exported = pieces[0];
                    local = pieces[0];
                } else if (pieces.length == 3 && "as".equals(pieces[1])) {
                    exported = pieces[0];
                    local = pieces[2];
                } else {
                    throw new IllegalArgumentException("Unsupported import binding: " + part);
                }
                out.append(" var ").append(local).append(" = ").append(moduleVar).append(".").append(exported).append(";");
            }
            return out.toString();
        }

        Matcher namespace = NAMESPACE_IMPORT.matcher(trimmed);
        if (namespace.matches()) {
            return leading + "var " + namespace.group(1) + " = require(\"" + namespace.group(2) + "\");";
        }

        Matcher sideEffect = SIDE_EFFECT_IMPORT.matcher(trimmed);
        if (sideEffect.matches()) {
            return leading + "require(\"" + sideEffect.group(1) + "\");";
        }

        throw new IllegalArgumentException("Unsupported import syntax: " + trimmed);
    }

    private static String leadingWhitespace(String value) {
        int i = 0;
        while (i < value.length() && (value.charAt(i) == ' ' || value.charAt(i) == '\t')) i += 1;
        return value.substring(0, i);
    }

    private static final class JsxParser {
        private final String source;
        private int index;

        JsxParser(String source) {
            this.source = source;
        }

        String transpile() {
            StringBuilder out = new StringBuilder(source.length() + 128);
            while (!isAtEnd()) {
                if (startsLineComment()) out.append(consumeLineComment());
                else if (startsBlockComment()) out.append(consumeBlockComment());
                else if (current() == '"' || current() == '\'' || current() == '`') out.append(consumeQuotedString(current()));
                else if (current() == '<' && looksLikeJsxElementStart(index)) out.append(parseElement());
                else out.append(advance());
            }
            return out.toString();
        }

        private String parseElement() {
            consume('<');
            String tag = parseTagName();
            if (tag.isEmpty()) throw new IllegalArgumentException("unterminated JSX element");

            List<String[]> attributes = new ArrayList<>();
            boolean selfClosing = false;
            while (!isAtEnd()) {
                skipWhitespace();
                if (startsWith("/>")) {
                    index += 2;
                    selfClosing = true;
                    break;
                }
                if (current() == '>') {
                    advance();
                    break;
                }
                attributes.add(parseAttribute());
            }

            List<String> children = new ArrayList<>();
            boolean closed = selfClosing;
            while (!selfClosing && !isAtEnd()) {
                if (startsWith("</")) {
                    index += 2;
                    String closingTag = parseTagName();
                    skipWhitespace();
                    consume('>');
                    if (!closingTag.equals(tag)) {
                        throw new IllegalArgumentException("mismatched JSX closing tag: expected </" + tag + ">, found </" + closingTag + ">");
                    }
                    closed = true;
                    break;
                }
                if (current() == '<' && looksLikeJsxElementStart(index)) {
                    children.add(parseElement());
                    continue;
                }
                if (current() == '{') {
                    String expression = parseBraceExpression();
                    if (!expression.trim().isEmpty()) children.add(expression);
                    continue;
                }
                String normalized = normalizeText(parseTextChild());
                if (!normalized.isEmpty()) children.add(jsStringLiteral(normalized));
            }
            if (!closed) throw new IllegalArgumentException("unterminated JSX element <" + tag + ">");

            List<String> args = new ArrayList<>();
            args.add(tagReference(tag));
            args.add(makeProps(attributes));
            args.addAll(children);
            return "JSX.h(" + String.join(", ", args) + ")";
        }

        private String[] parseAttribute() {
            String name = parseAttributeName();
            if (name.isEmpty()) throw new IllegalArgumentException("unsupported JSX attribute near " + current());
            skipWhitespace();
            if (current() != '=') return new String[] { name, "true" };
            advance();
            skipWhitespace();
            if (current() == '{') return new String[] { name, parseBraceExpression() };
            if (current() == '"' || current() == '\'') return new String[] { name, consumeQuotedString(current()) };
            int start = index;
            while (!isAtEnd() && !Character.isWhitespace(current()) && current() != '>' && !startsWith("/>")) advance();
            return new String[] { name, source.substring(start, index) };
        }

        private String parseBraceExpression() {
            consume('{');
            int start = index;
            int depth = 1;
            while (!isAtEnd()) {
                if (startsLineComment()) {
                    consumeLineComment();
                    continue;
                }
                if (startsBlockComment()) {
                    consumeBlockComment();
                    continue;
                }
                if (current() == '"' || current() == '\'' || current() == '`') {
                    consumeQuotedString(current());
                    continue;
                }
                char c = advance();
                if (c == '{') depth += 1;
                else if (c == '}') {
                    depth -= 1;
                    if (depth == 0) return source.substring(start, index - 1);
                }
            }
            throw new IllegalArgumentException("unterminated JSX expression");
        }

        private String parseTextChild() {
            int start = index;
            while (!isAtEnd() && current() != '<' && current() != '{') advance();
            return source.substring(start, index);
        }

        private String parseTagName() {
            int start = index;
            while (!isAtEnd() && isTagNameCharacter(current())) advance();
            return source.substring(start, index);
        }

        private String parseAttributeName() {
            int start = index;
            while (!isAtEnd() && isAttributeNameCharacter(current())) advance();
            return source.substring(start, index);
        }

        private String makeProps(List<String[]> attributes) {
            if (attributes.isEmpty()) return "null";
            List<String> pairs = new ArrayList<>();
            for (String[] attribute : attributes) pairs.add(propertyKey(attribute[0]) + ": " + attribute[1]);
            return "{ " + String.join(", ", pairs) + " }";
        }

        private String propertyKey(String name) {
            return isIdentifier(name) ? name : jsStringLiteral(name);
        }

        private String tagReference(String tag) {
            return isIdentifier(tag) ? tag : jsStringLiteral(tag);
        }

        private String normalizeText(String text) {
            return text.trim().replaceAll("\\s+", " ");
        }

        private String jsStringLiteral(String value) {
            StringBuilder out = new StringBuilder("\"");
            for (int i = 0; i < value.length(); i++) {
                char c = value.charAt(i);
                switch (c) {
                    case '"': out.append("\\\""); break;
                    case '\\': out.append("\\\\"); break;
                    case '\n': out.append("\\n"); break;
                    case '\r': out.append("\\r"); break;
                    case '\t': out.append("\\t"); break;
                    default:
                        if (c < 0x20) out.append(String.format(Locale.US, "\\u%04X", (int) c));
                        else out.append(c);
                }
            }
            return out.append('"').toString();
        }

        private boolean isAtEnd() { return index >= source.length(); }
        private char current() { return isAtEnd() ? '\0' : source.charAt(index); }
        private char peek() { int next = index + 1; return next < source.length() ? source.charAt(next) : '\0'; }
        private char advance() { return source.charAt(index++); }
        private boolean startsLineComment() { return current() == '/' && peek() == '/'; }
        private boolean startsBlockComment() { return current() == '/' && peek() == '*'; }
        private boolean startsWith(String text) { return source.startsWith(text, index); }
        private void skipWhitespace() { while (!isAtEnd() && Character.isWhitespace(current())) advance(); }
        private void consume(char expected) {
            if (current() != expected) throw new IllegalArgumentException("expected " + expected);
            advance();
        }

        private String consumeLineComment() {
            int start = index;
            while (!isAtEnd() && current() != '\n') advance();
            if (!isAtEnd()) advance();
            return source.substring(start, index);
        }

        private String consumeBlockComment() {
            int start = index;
            index += 2;
            while (!isAtEnd()) {
                if (current() == '*' && peek() == '/') {
                    index += 2;
                    break;
                }
                advance();
            }
            return source.substring(start, index);
        }

        private String consumeQuotedString(char quote) {
            int start = index;
            advance();
            while (!isAtEnd()) {
                char c = advance();
                if (c == '\\') {
                    if (!isAtEnd()) advance();
                    continue;
                }
                if (c == quote) break;
            }
            return source.substring(start, index);
        }

        private boolean looksLikeJsxElementStart(int start) {
            if (start >= source.length() || source.charAt(start) != '<') return false;
            int cursor = start + 1;
            if (cursor >= source.length() || !Character.isUpperCase(source.charAt(cursor))) return false;
            while (cursor < source.length() && isTagNameCharacter(source.charAt(cursor))) cursor += 1;
            while (cursor < source.length() && Character.isWhitespace(source.charAt(cursor))) cursor += 1;
            if (cursor >= source.length()) return false;
            char c = source.charAt(cursor);
            if (c == '>') return true;
            if (c == '/') return cursor + 1 < source.length() && source.charAt(cursor + 1) == '>';
            return isAttributeNameStart(c);
        }

        private boolean isTagNameCharacter(char c) { return Character.isLetterOrDigit(c) || c == '_' || c == '.'; }
        private boolean isAttributeNameCharacter(char c) { return Character.isLetterOrDigit(c) || c == '_' || c == '-' || c == ':'; }
        private boolean isAttributeNameStart(char c) { return Character.isLetter(c) || c == '_' || c == ':'; }
        private boolean isIdentifier(String value) {
            if (value == null || value.isEmpty()) return false;
            char first = value.charAt(0);
            if (!Character.isLetter(first) && first != '_' && first != '$') return false;
            for (int i = 1; i < value.length(); i++) {
                char c = value.charAt(i);
                if (!Character.isLetterOrDigit(c) && c != '_' && c != '$') return false;
            }
            return true;
        }
    }
}
